### ISSUE TITLE
Another module header bug

### ISSUE THREAD

**Issue by snoyberg** [fpco/ide-backend#92]
This issue is closely related to #91, but seems to be fairly different. Most importantly, while I can write a hack to work around #91, I do not believe any hack will help here. Consider the following code:

``` haskell
{-# LANGUAGE OverloadedStrings #-}

module Main where

import IdeSession
import Data.Monoid (mconcat)

main :: IO ()
main = do
    sess <- initSession defaultSessionConfig
                { configDir             = "."
                }

    let cb = \_ -> return ()
        update = flip (updateSession sess) cb

    update $ mconcat
        [ updateModule "Main.hs" "module Main where\nimport Something"
        , updateModule "Something2.hs" "module Something where"
        ]

    getSourceErrors sess >>= print

    update $ mconcat
        [ updateModule "Main.hs" "module Main where\nimport Something"
        , updateModule "Something2.hs" "module Something2 where"
        ]

    getSourceErrors sess >>= print
```

Both versions of the code which have been updated are incorrect. In the first case, we have an incorrect module header for the named file. This incorrect code is _not_ flagged by ide-backend. This is the issue described in #91.

However, the second code update fixes this issue, and `Something2.hs` is now a perfectly valid module by all counts. `Main`, however, is invalid, since it imports the module `Something` which does not exist. However, the error message generated by ide-backend is:

```
[SourceError {errorKind = KindError, errorSpan = Something2.hs@1:8-1:18, errorMsg = "File name does not match module name:\nSaw: `Something2'\nExpected: `Something'\n"}]
```

In other words, ide-backend is flagging the wrong module as being buggy, with an incredibly confusing error message, since there's no reason why `Something2.hs` should be expected to have a module named `Something`.

Pinging @dcoutts @Mikolaj @edsko @mgsloan @jwiegley
